### PR TITLE
Refactor network policy e2e tests, and add additional logging when tests fail

### PR DESCRIPTION
**Pull request by gunjan5**
**What this PR does / why we need it**:
- Restructure tests, clean up duplicate code
- Add additional logging, so it's easier to debug when the tests fail
- Add retry on the client side pods because sometimes kubelet will start the pod when the network is not ready, resulting in flaky tests
- Add readiness probe for the server pods to further reduce the flakiness.

Current failure messages when a test fails:
```
checking client-can-connect could communicate with server.
  Expected error:
      <*errors.errorString | 0xc420868820>: {
          s: "pod \"client-can-connect\" failed with status: {Phase:Failed Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-07-25 15:54:48 -0700 PDT Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-07-25 15:54:57 -0700 PDT Reason:ContainersNotReady Message:containers with unready status: [client-can-connect-container]} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-07-25 15:54:48 -0700 PDT Reason: Message:}] Message: Reason: HostIP:10.138.0.2 PodIP:10.60.1.180 StartTime:2017-07-25 15:54:48 -0700 PDT InitContainerStatuses:[] ContainerStatuses:[{Name:client-can-connect-container State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:1,Signal:0,Reason:Error,Message:,StartedAt:2017-07-25 15:54:49 -0700 PDT,FinishedAt:2017-07-25 15:54:57 -0700 PDT,ContainerID:docker://c91c33139672c20a5231bcba9437d74e12d35a6d19275b41bb8e571148b4ab57,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:gcr.io/google_containers/redis:e2e ImageID:docker://sha256:e5e67996c442f903cda78dd983ea6e94bb4e542950fd2eba666b44cbd303df42 ContainerID:docker://c91c33139672c20a5231bcba9437d74e12d35a6d19275b41bb8e571148b4ab57}] QOSClass:BestEffort}",
      }
      pod "client-can-connect" failed with status: {Phase:Failed Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-07-25 15:54:48 -0700 PDT Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-07-25 15:54:57 -0700 PDT Reason:ContainersNotReady Message:containers with unready status: [client-can-connect-container]} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-07-25 15:54:48 -0700 PDT Reason: Message:}] Message: Reason: HostIP:10.138.0.2 PodIP:10.60.1.180 StartTime:2017-07-25 15:54:48 -0700 PDT InitContainerStatuses:[] ContainerStatuses:[{Name:client-can-connect-container State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:1,Signal:0,Reason:Error,Message:,StartedAt:2017-07-25 15:54:49 -0700 PDT,FinishedAt:2017-07-25 15:54:57 -0700 PDT,ContainerID:docker://c91c33139672c20a5231bcba9437d74e12d35a6d19275b41bb8e571148b4ab57,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:gcr.io/google_containers/redis:e2e ImageID:docker://sha256:e5e67996c442f903cda78dd983ea6e94bb4e542950fd2eba666b44cbd303df42 ContainerID:docker://c91c33139672c20a5231bcba9437d74e12d35a6d19275b41bb8e571148b4ab57}] QOSClass:BestEffort}
  not to have occurred
```

Failure message with this PR:
```
 should support a 'default-deny' policy [Feature:NetworkPolicy] [It]
    /Users/gunjan/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/network_policy.go:81

    Jul 25 14:35:55.978: Pod client-cannot-connect should be able to connect to service svc-server, but was not able to connect.
    Pod logs:
    Connecting to svc-server.e2e-tests-network-policy-tbzm8:80 (10.63.254.124:80)
    wget: download timed out
    Connecting to svc-server.e2e-tests-network-policy-tbzm8:80 (10.63.254.124:80)
    wget: download timed out
    Connecting to svc-server.e2e-tests-network-policy-tbzm8:80 (10.63.254.124:80)
    wget: download timed out
    Connecting to svc-server.e2e-tests-network-policy-tbzm8:80 (10.63.254.124:80)
    wget: download timed out
    Connecting to svc-server.e2e-tests-network-policy-tbzm8:80 (10.63.254.124:80)
    wget: download timed out


   Current NetworkPolicies:
    	[{{ } {deny-all  e2e-tests-network-policy-tbzm8 /apis/networking.k8s.io/v1/namespaces/e2e-tests-network-policy-tbzm8/networkpolicies/deny-all 20fa92ab-7181-11e7-855b-42010a8a0003 87694 1 2017-07-25 14:35:07 -0700 PDT <nil> <nil> map[] map[] [] nil [] } {{map[] []} []}}]

   Pods:
    	[Pod: server, Status: &PodStatus{Phase:Running,Conditions:[{Initialized True 0001-01-01 00:00:00 +0000 UTC 2017-07-25 14:34:52 -0700 PDT  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2017-07-25 14:35:00 -0700 PDT  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2017-07-25 14:34:52 -0700 PDT  }],Message:,Reason:,HostIP:10.138.0.2,PodIP:10.60.1.145,StartTime:2017-07-25 14:34:52 -0700 PDT,ContainerStatuses:[{server-container-80 {nil ContainerStateRunning{StartedAt:2017-07-25 14:34:53 -0700 PDT,} nil} {nil nil nil} true 0 gcr.io/google_containers/porter:4524579c0eb935c056c8e75563b4e1eda31587e0 docker://sha256:711ccbc1c74f3417f2a0f403c194017b3bde09eec3c3b42dc9c838db3894c223 docker://3fa73da57380ef28520f8bbf53b3540f523ab76d45a1b42cc9a1e4573eb45639}],QOSClass:BestEffort,InitContainerStatuses:[],}
     Pod: server, Status: &PodStatus{Phase:Running,Conditions:[{Initialized True 0001-01-01 00:00:00 +0000 UTC 2017-07-25 14:34:52 -0700 PDT  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2017-07-25 14:35:00 -0700 PDT  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2017-07-25 14:34:52 -0700 PDT  }],Message:,Reason:,HostIP:10.138.0.2,PodIP:10.60.1.145,StartTime:2017-07-25 14:34:52 -0700 PDT,ContainerStatuses:[{server-container-80 {nil ContainerStateRunning{StartedAt:2017-07-25 14:34:53 -0700 PDT,} nil} {nil nil nil} true 0 gcr.io/google_containers/porter:4524579c0eb935c056c8e75563b4e1eda31587e0 docker://sha256:711ccbc1c74f3417f2a0f403c194017b3bde09eec3c3b42dc9c838db3894c223 docker://3fa73da57380ef28520f8bbf53b3540f523ab76d45a1b42cc9a1e4573eb45639}],QOSClass:BestEffort,InitContainerStatuses:[],}
    ]
```
**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by caseydavenport** [comment]
CC @kubernetes/sig-network-pr-reviews @danwinship

File: test/e2e/network_policy.go [hunk 1]
```diff
@@ -418,13 +434,10 @@ func createNetworkClientPod(f *framework.Framework, namespace *v1.Namespace, pod
...
 					Name:  fmt.Sprintf("%s-container", podName),
-					Image: "gcr.io/google_containers/busybox:1.24",
+					Image: "gcr.io/google_containers/alpine-with-bash:1.0",
 					Args: []string{
-						"/bin/wget",
-						"-T", "8",
-						fmt.Sprintf("%s.%s:%d", targetService.Name, targetService.Namespace, targetPort),
-						"-O",
-						"-",
+						fmt.Sprintf("for i in {1..5}; do wget -T 8 %s.%s:%d -O - && exit 0 || sleep 1; done; exit 1",
```
Comment (danwinship) [inline test/e2e/network_policy.go]: You've added a readiness probe to the server pod now, so we should just not run the client pod until the server pod is ready.

File: test/e2e/network_policy.go [hunk 2]
```diff
@@ -39,286 +39,255 @@ failed or succesfully connected as expected.
...
-		policy := &networking.NetworkPolicy{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: "deny-all",
-			},
-			Spec: networking.NetworkPolicySpec{
-				PodSelector: metav1.LabelSelector{},
-				Ingress:     []networking.NetworkPolicyIngressRule{},
-			},
-		}
+	Context("Single-port servers", func() {
```
Comment (danwinship) [inline test/e2e/network_policy.go]: Reorg looks generally good; I didn't look in extreme detail.

Do you really need the single-port servers/multi-port servers split? Couldn't you just always listen on both ports but just ignore port 81 in the tests where you don't care about it?

File: test/e2e/network_policy.go [hunk 3]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package e2e
+package network
```
Comment (danwinship) [inline test/e2e/network_policy.go]: what is this for?

File: test/e2e/network_policy.go [hunk 4]
```diff
@@ -75,9 +75,9 @@ var _ = framework.KubeDescribe("NetworkPolicy", func() {
 			Expect(err).NotTo(HaveOccurred())
 			defer cleanupNetworkPolicy(f, policy)
 
-			// Create a pod with name 'client-cannot-connect', which will attempt to comunicate with the server,
+			// Create a pod with name 'client-cannot-connect', which will attempt to communicate with the server,
 			// but should not be able to now that isolation is on.
-			testCanConnect(f, f.Namespace, "client-cannot-connect", service, 80)
+			testCannotConnect(f, f.Namespace, "client-cannot-connect", service, 80)
```
Comment (danwinship) [inline test/e2e/network_policy.go]: this is a fix to a previous commit, it shouldn't be in in this commit

File: test/e2e/network_policy.go [hunk 5]
```diff
@@ -338,7 +307,13 @@ func testCanConnect(f *framework.Framework, ns *v1.Namespace, podName string, se
 
 	framework.Logf("Waiting for %s to complete.", podClient.Name)
 	err = framework.WaitForPodSuccessInNamespace(f.ClientSet, podClient.Name, ns.Name)
-	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("checking %s could communicate with server.", podClient.Name))
+	if err != nil {
+		logs, logErr := framework.GetPodLogs(f.ClientSet, f.Namespace.Name, podName, fmt.Sprintf("%s-container", podName))
```
Comment (danwinship) [inline test/e2e/network_policy.go]: should be in the "add additional logs when tests fail" commit?

**Review by danwinship — CHANGES_REQUESTED** [review]
The second "add additional logs when tests fail" commit seems to be the result of accidentally merging in another branch or something? That needs to be fixed. A few other comments below

File: test/e2e/network_policy.go (hunk 4 above)
Comment (gunjan5) [inline test/e2e/network_policy.go]: I'm planning to squash all the commits into one

File: test/e2e/network_policy.go (hunk 2 above)
Comment (gunjan5) [inline test/e2e/network_policy.go]: good point

**Comment by gunjan5** [comment]
@danwinship apologize for the delay. I've addressed your comments and squashed the commits. PTAL

File: test/e2e/network/network_policy.go [hunk 6]
```diff
@@ -36,290 +38,247 @@ are correctly enforced by a networking plugin. It accomplishes this by launching
...
+		BeforeEach(func() {
+			By("Creating a simple server that serves on port 80 and 81.")
+			podServer, service = createServerPodAndService(f, f.Namespace, "server", []int{80, 81})
+			framework.Logf("Waiting for Server to come up.")
+			err := framework.WaitForPodRunningInNamespace(f.ClientSet, podServer)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Wait for the server pods to become ready.
+			for !(podServer.Status.ContainerStatuses[0].Ready && podServer.Status.ContainerStatuses[1].Ready) {
+				time.Sleep(1 * time.Second)
```
Comment (danwinship) [inline test/e2e/network/network_policy.go]: This won't work; you need to re-fetch the Pod object from the server. The existing Pod object that you have won't change.

But I'm pretty sure there are already "wait for a pod to be ready" utility functions somewhere in test/e2e

File: test/e2e/network/network_policy.go [hunk 7]
```diff
@@ -448,11 +478,12 @@ func createNetworkClientPod(f *framework.Framework, namespace *v1.Namespace, pod
 			Containers: []v1.Container{
 				{
 					Name:  fmt.Sprintf("%s-container", podName),
-					Image: imageutils.GetE2EImage(imageutils.Redis),
+					Image: imageutils.GetBusyBoxImage(),
 					Args: []string{
 						"/bin/sh",
 						"-c",
-						fmt.Sprintf("/usr/bin/printf dummy-data | /bin/nc -w 8 %s.%s %d", targetService.Name, targetService.Namespace, targetPort),
+						fmt.Sprintf("for i in $(seq 1 5); do wget -T 8 %s.%s:%d -O - && exit 0 || sleep 1; done; exit 1",
```
Comment (danwinship) [inline test/e2e/network/network_policy.go]: again, as long as the pod is ready, and the service has an endpoint, then we shouldn't need to retry; the first attempt should succeed

**Review by danwinship — CHANGES_REQUESTED** [review]

**Comment by caseydavenport** [comment]
@gunjan5 @bcreane this generally LGTM.  I'd like to give @danwinship another chance to review if he's interested.

**Comment by danwinship** [comment]
> again, as long as the pod is ready, and the service has an endpoint, then we shouldn't need to retry; the first attempt should succeed

Hm... I had said this, but now looking at the code, there's nothing that ensures the service has an endpoint yet. (Maybe the old code did that via one of the helper functions it used, and the new code doesn't?)

So anyway, either call something to make sure that the service endpoint has been created as part of the setup, or else put back the retrying. Sorry about that.

**Comment by bcreane** [comment]
@danwinship - thanks for your reviews and spotting the missing service endpoint readiness check. I've restored the original `wget` probe since I don't see a straight forward way to do this with the framework. The PR should be updated shortly.

**Comment by bcreane** [comment]
@danwinship - PR updated, PTAL.

**Review by danwinship — APPROVED** [review]

**Pull request merged**